Basic implementation of native schema string indexing layer

tinwelint opened the pull request:
Much of the changes in here are about making test classes and utilities
generic to value type as well as copy type-specific indexing code from
number schema index classes.

Here's also a basic, non-optimized implementation of the string schema
index accessor, string layout and friends which focuses on getting the
functionality in place first and foremost. This string index isn't hooked in
at any place, only some tests assert its functionality.

There are still some TODOs scattered around the code, which are mostly 
points to attack next or to consider more carefully when moving into
performance and integration work, please don't take too much notice
of them as they are here mostly for reminders when doing further development.

co-author: @burqen

tinwelint committed:
Basic implementation of native schema string indexing layer

Much of the changes in here are about making test classes and utilities
generic to value type as well as copy type-specific indexing code from
number schema index classes.

Here's also a basic, non-optimized implementation of the string schema
index accessor, string layout and friends which focuses on getting the
functionality in place first and foremost.

tinwelint committed:
String IndexPopulator tests working

tinwelint committed:
Implemented StringSchemaIndexProvider

Broke out commonlatities from NumberSchemaIndexProvider into NativeSchemaIndexProvider

tinwelint committed:
Improved on some javadoc and cut down on unnecessary generics

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/StringLayout.java:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    @Override
+    public StringSchemaKey copyKey( StringSchemaKey key, StringSchemaKey into )
```
can you please describe when it should be used (like node splitting) because its not clear why its used for example during the consistency check

burqen replied:
The `ConsistencyChecker` that is using it is actually the consistency checker for the `GBPTree` itself, not the Neo consistency checker. You use it when you want a copy of the key that you can make changes to without destroying the original state.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/StringSchemaKey.java:
```diff
@@ -0,0 +1,226 @@
+    static final int ENTITY_ID_SIZE = Long.BYTES;
+
+    private long entityId;
+    private boolean entityIdIsSpecialTieBreaker;
```
what is this field about?

burqen replied:
Keys look like `<stringValue,nodeId>`. We want to do a range scan over `"A"-"C"` _inclusive_ in both ends. We initiate our keys that define our range as `<"A",Long.MIN_VALUE> - <"C",Long.MAX_VALUE>` which will cover the entire range of possible values. The problem is that if do not explicitly take the node id part into account we will only compare against the string value in an exclusive fashion. So to tell the comparator that he needs to take the id into consideration as well we do `entityIdIsSpecialTieBreaker=true`.

MishaDemianenko replied:
then why not call this field `compareIds` or something that will not force future reader to break his tie?

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/StringLayoutNonUnique.java:
```diff
@@ -0,0 +1,58 @@
+public class StringLayoutNonUnique extends StringLayout
+{
+    private static final String IDENTIFIER_NAME = "NUSI";
+    static final int MAJOR_VERSION = 0;
```
all constants here can be private

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/StringLayoutUnique.java:
```diff
@@ -0,0 +1,66 @@
+class StringLayoutUnique extends StringLayout
+{
+    private static final String IDENTIFIER_NAME = "USI";
+    static final int MAJOR_VERSION = 0;
```
all fields here can be private

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/StringSchemaIndexProvider.java:
```diff
@@ -0,0 +1,120 @@
+{
+    public static final String KEY = "string";
+    public static final Descriptor STRING_PROVIDER_DESCRIPTOR = new Descriptor( KEY, "1.0" );
+    static final IndexCapability CAPABILITY = new StringIndexCapability();
```
CAPABILITY field should be private

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/StringSchemaIndexProvider.java:
```diff
@@ -0,0 +1,120 @@
+public class StringSchemaIndexProvider extends NativeSchemaIndexProvider<StringSchemaKey,NativeSchemaValue>
+{
+    public static final String KEY = "string";
+    public static final Descriptor STRING_PROVIDER_DESCRIPTOR = new Descriptor( KEY, "1.0" );
```
As I can see we have versions for layout and for the provider, how would potential future supported migration paths look like?

burqen replied:
Keep `native-lucene-1.0` around as is.
Introduce new `native-lucene-2.0` that include native string.
Do no automatic migration and instead let user use drop+create just as when we introduced native number.

So far we have not had the need to utilise the layout version but it's their so that we _can_ if we need in the future. One example would be if we wound move native number layout from "fixed" to "dynamic". The migration path in that case is not super clear.

MishaDemianenko replied:
talked offline

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/StringSchemaIndexProvider.java:
```diff
@@ -0,0 +1,120 @@
+        return CAPABILITY;
+    }
+
+    private static class StringIndexCapability implements IndexCapability
```
please add a javadoc that describe underlying capabilities

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/StringSchemaIndexReader.java:
```diff
@@ -0,0 +1,117 @@
+
+        if ( indexOrder != IndexOrder.NONE )
+        {
+            throw new UnsupportedOperationException( "unsupported order " + indexOrder );
```
In StringCapabilities we claim that we support `ASCENDING` index order. Why do we not allow it here?, and shouldn't we use capabilities here for validation of predicates as well?

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/StringSchemaIndexReader.java:
```diff
@@ -0,0 +1,117 @@
+            break;
+        case exact:
+            ExactPredicate exactPredicate = (ExactPredicate) predicate;
+            treeKeyFrom.from( Long.MIN_VALUE, exactPredicate.value() );
```
i find following predicates too low level at this point: key.initAsLowest and key.initAsHighest are really good examples how keys should be initialized on this level and I would prefer to have helper methods instead that will perform initialisations for exact and range cases as well

MishaDemianenko replied:
well, initFromForRange and initFromForRange is still here

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/StringSchemaKey.java:
```diff
@@ -0,0 +1,226 @@
+
+    private TextValue assertValidValue( Value... values )
+    {
+        if ( values.length > 1 )
```
this more or less duplicate capability checks, maybe use that instead?

burqen replied:
Yes, the validation of "what goes where" happens all over the place right now. Created a card for cleaning this up. Ok if we leave as future work?

MishaDemianenko replied:
oke

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/StringSchemaKey.java:
```diff
@@ -0,0 +1,226 @@
+    int compareValueTo( StringSchemaKey other )
+    {
+        // TODO cover all cases of bytes == null and special tie breaker and document
+        if ( bytes != other.bytes )
```
this is wrong: for example, comparing highest and lowest key now will say they are the same

burqen replied:
It's actually ok because we want to `compareValueInThisKey` and not `compareKey`. Validation of highest and lowest is done outside of this method.

MishaDemianenko replied:
your future reader would be glad to know that there are specific values that considered outside of this method and also be curious why its so

burqen replied:
Just like the current readers then ;) Do you think javadoc would be enough?

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/StringSchemaKey.java:
```diff
@@ -0,0 +1,226 @@
+            }
+        }
+
+        if ( a.length < b.length )
```
Integer.compare( a.length, b.length )

MishaDemianenko replied:
All of the lines below the first comparison can be replaced with only one integer.compare thingy

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/StringSchemaKey.java:
```diff
@@ -0,0 +1,226 @@
+        }
+        catch ( Exception e )
+        {
+            // We can not throw here because we will visit this method inside a pageCursor.shouldRetry() block.
```
not sure if its safe to do a fallback here since it will mean that comparison order is wrong anyway

burqen replied:
This will be re-written in future work.

MishaDemianenko commented on community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/NativeSchemaIndexProviderTest.java:
```diff
@@ -0,0 +1,422 @@
+        }
+    }
+
+    protected abstract Value someValue();
```
nice to see this hidden abstract method between the tests, otherwise its to easy to find all f those abstract things...

MishaDemianenko commented on community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/NumberSchemaIndexAccessorTest.java:
```diff
@@ -81,4 +82,28 @@ public void respectIndexOrder() throws Exception
             assertTrue( "found all values", i == expectedValues.length );
```
assertEquals?

MishaDemianenko reviewed (changes requested):
Couple of comments here and there, have tried to avoid commenting todos here and there as noted in description

Commit:
Review cleanup 1

Commit:
CpabilityValidator introduced to validate queries

Commit:
Review cleanup 2

Commit:
Safe StringLayout.readKey

burqen commented:
@MishaDemianenko Comments addressed and rebase done :D

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/StringSchemaKey.java:
```diff
@@ -0,0 +1,226 @@
+            }
+        }
+
+        if ( Integer.compare( a.length, b.length ) < 0 )
```
_whole_ thing below this line can be replaced with single Integer.compare

burqen replied:
Haha, yes ofc! xD

Commit:
Review cleanup

- Change field name 'entityIdIsSpecialTieBreaker' to 'compareId'
- Replace if-else with Integer.compare

burqen commented:
@MishaDemianenko One more try

Commit:
Cleanup after rebase

Commit:
Code cleanup around schema
